### PR TITLE
Add checks to roi_heads in detection module

### PR DESCRIPTION
**Pull request by ekagra-ranjan**
This PR addresses #1027 by:
1)  adding checks on `boxes`, `masks`, `keypoints` and `labels` attributes of target passed to detection models. 
2) updating docs

### CODE REVIEW

File: torchvision/models/detection/roi_heads.py [hunk 1]
```diff
@@ -525,6 +527,15 @@ def forward(self, features, proposals, image_shapes, targets=None):
             image_shapes (List[Tuple[H, W]])
             targets (List[Dict])
         """
+        if targets is not None:
+            for t in targets:
+                assert t["boxes"].dtype == torch.float32, 'target boxes must of float type'
```
Comment (fmassa) [inline torchvision/models/detection/roi_heads.py]: Those checks might not work if we want to perform fp16 training in the future.
Maybe might be better to use something like `t["boxes"].dtype.is_floating_point`?

File: torchvision/models/detection/roi_heads.py [hunk 2]
```diff
@@ -525,6 +527,15 @@ def forward(self, features, proposals, image_shapes, targets=None):
...
+                assert t["labels"].dtype == torch.int64, 'target labels must of int64 type'
+                if self.has_mask:
+                    assert t["masks"].dtype == torch.uint8, 'target masks must of uint8 type'
```
Comment (fmassa) [inline torchvision/models/detection/roi_heads.py]: Do the masks *need* to be `uint8`? I thought that the code worked as well for other types, as the only place where it is used in the model we perform a cast to `float`?

File: torchvision/models/detection/faster_rcnn.py [hunk 3]
```diff
@@ -30,21 +30,21 @@ class FasterRCNN(GeneralizedRCNN):
 
     The behavior of the model changes depending if it is in training or evaluation mode.
 
-    During training, the model expects both the input tensors, as well as a targets dictionary,
+    During training, the model expects both the input tensors, as well as a targets (list of dictionary),
     containing:
-        - boxes (Tensor[N, 4]): the ground-truth boxes in [x0, y0, x1, y1] format, with values
+        - boxes (FloatTensor[N, 4]): the ground-truth boxes in [x0, y0, x1, y1] format, with values
```
Comment (fmassa) [inline torchvision/models/detection/faster_rcnn.py]: Those are great changes, thanks!

But I'm a bit concerned that this looks like the legacy interface of `torch.FloatTensor`, etc.
I wonder if there is a better way of representing this? For example, in numpy, everything is a `ndarray`, but with different types.

Thoughts?

File: torchvision/models/detection/roi_heads.py (hunk 1 above)
Comment (ekagra-ranjan) [reply torchvision/models/detection/roi_heads.py]: Yes, you are right! Will change it.

File: torchvision/models/detection/roi_heads.py (hunk 2 above)
Comment (ekagra-ranjan) [reply torchvision/models/detection/roi_heads.py]: I was following the [tutorial](https://colab.research.google.com/github/pytorch/vision/blob/temp-tutorial/tutorials/torchvision_finetuning_instance_segmentation.ipynb), where the dtype of mask was specified `uint8`. Should I remove that check?

File: torchvision/models/detection/faster_rcnn.py (hunk 3 above)
Comment (ekagra-ranjan) [reply torchvision/models/detection/faster_rcnn.py]: Would the descriptions like `boxes(Tensor[N, 4], dtype=torch.float)` or `boxes(FloatTensor(N, 4))` be better? 

`(Tensor[N, 4], dtype=torch.float)` is not a valid syntax but conveys the requirement whereas `FloatTensor(N, 4)` would actually create a float tensor with `dim = (N, 4)`.

Do let me know if you have any better way of representing them?

**Comment by ekagra-ranjan** [comment]
Sorry for the delayed response. Please have a look at my response @fmassa .

File: torchvision/models/detection/roi_heads.py (hunk 2 above)
Comment (fmassa) [reply torchvision/models/detection/roi_heads.py]: I don't think having the masks being `uint8` is a hard-restriction, so yes, maybe remove this check for now

File: torchvision/models/detection/faster_rcnn.py (hunk 3 above)
Comment (fmassa) [reply torchvision/models/detection/faster_rcnn.py]: Let's just keep this as is for now. But I'd like to remove the `FloatTensor`, changes, and only keep `Int64Tensor` instead, because boxes is not required to be `Float`

File: torchvision/models/detection/roi_heads.py (hunk 2 above)
Comment (ekagra-ranjan) [reply torchvision/models/detection/roi_heads.py]: Done.

**Comment by ekagra-ranjan** [comment]
@fmassa Made the changes!
